File: app/src/main/java/io/mgba/Adapters/Settings/SettingsAdapter.java
package io.mgba.Adapters.Settings;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.mgba.Adapters.Interfaces.BaseAdapter;
import io.mgba.Adapters.Interfaces.SingleViewHolderAdapter;
import io.mgba.Data.Settings.Settings;
import io.mgba.R;
import io.mgba.mgba;
import io.reactivex.functions.Consumer;

public class SettingsAdapter extends SingleViewHolderAdapter<Settings> {
    private static final String TAG = "mgba:SettingsAdapter";
    private final Consumer<Settings> onClick;

    public SettingsAdapter(List<Settings> settings, Context context, Consumer<Settings> onClick, RecyclerView recyclerView) {
        super(settings, recyclerView, context, R.layout.category_element, ViewHolder::new);
        this.onClick = onClick;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Settings item = items.get(position);

        ((ViewHolder) holder).icon.setImageDrawable(mCtx.getDrawable(item.getResource()));
        ((ViewHolder) holder).title.setText(item.getTitle());

        ((ViewHolder) holder).container.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        final Settings mItem = items.get(getPositionBasedOnView(v));

        try {
            onClick.accept(mItem);
        } catch (Exception e) {
            mgba.printLog(TAG, "Cannot execute consumer callable");
        }
    }

    public static class ViewHolder extends RecyclerView.ViewHolder{
        @BindView(R.id.setting_icon)
        ImageView icon;
        @BindView(R.id.setting_title)
        TextView title;
        @BindView(R.id.container)
        RelativeLayout container;

        public ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }
}

File: app/src/main/java/io/mgba/adapters/SettingsAdapter.kt
package io.mgba.adapters

import android.content.Context
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import io.mgba.adapters.interfaces.SingleViewHolderAdapter
import io.mgba.data.settings.Settings
import io.mgba.R
import kotlinx.android.synthetic.main.category_element.view.*

class SettingsAdapter(settings: List<Settings>, context: Context, private val onClick: (Settings) -> Any, recyclerView: RecyclerView) : SingleViewHolderAdapter<Settings>(settings, recyclerView, context, R.layout.category_element, {v -> ViewHolder(v) }) {

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = items[position]

        (holder as ViewHolder).icon!!.setImageDrawable(mCtx.getDrawable(item.resource))
        holder.title!!.text = item.title

        holder.container!!.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        val mItem = items[getPositionBasedOnView(v)]
        onClick.invoke(mItem)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        internal var icon = view.setting_icon
        internal var title = view.setting_title
        internal var container = view.container
    }

    companion object {
        private val TAG = "mgba:SettingsAdapter"
    }
}
